pennylane_lightning/src/util/Util.hpp

```diff
@@ -204,89 +207,465 @@ template <class T> inline auto dimSize(const std::vector<T> &data) -> size_t {
     return static_cast<size_t>(log2(s_sqrt));
 }
 
+/**
+ * @brief Calculates the inner-product using OpenMP.
+ *
+ * @tparam T Floating point precision type.
+ * @param v1 Complex data array 1.
+ * @param v2 Complex data array 2.
+ * @param result Calculated inner-product of v1 and v2.
+ * @param data_size Size of data arrays.
+ */
+template <class T>
+inline static void
+omp_innerProd(const std::complex<T> *v1, const std::complex<T> *v2,
+              std::complex<T> &result, const size_t data_size) {
+#if defined _OPENMP
+#pragma omp declare \
+            reduction (sm:std::complex<T>:omp_out=ConstSum(omp_out, omp_in)) \
+            initializer(omp_priv=std::complex<T> {0, 0})
+#endif
+
+#if defined _OPENMP
+#pragma omp parallel for default(none) shared(v1, v2, data_size)               \
+    reduction(sm                                                               \
+              : result)
+#endif
+    for (size_t i = 0; i < data_size; i++) {
+        result = ConstSum(result, ConstMult(*(v1 + i), *(v2 + i)));
+    }
+}
+
 /**
  * @brief Calculates the inner-product using the best available method.
  *
  * @tparam T Floating point precision type.
- * @param data_1 Complex data array 1.
- * @param data_2 Complex data array 2.
+ * @param v1 Complex data array 1.
+ * @param v2 Complex data array 2.
  * @param data_size Size of data arrays.
  * @return std::complex<T> Result of inner product operation.
  */
 template <class T>
-auto innerProd(const std::complex<T> *data_1, const std::complex<T> *data_2,
-               const size_t data_size) -> std::complex<T> {
+inline auto innerProd(const std::complex<T> *v1, const std::complex<T> *v2,
+                      const size_t data_size) -> std::complex<T> {
     std::complex<T> result(0, 0);
 
-    if constexpr (USE_CBLAS) {
-        if constexpr (std::is_same_v<T, float>) {
-            cblas_cdotu_sub(data_size, data_1, 1, data_2, 1, &result);
-        } else if constexpr (std::is_same_v<T, double>) {
-            cblas_zdotu_sub(data_size, data_1, 1, data_2, 1, &result);
-        }
-    } else {
+#if defined(USE_CBLAS) && USE_CBLAS
```

Comment:
Does `if constexpr` not work here? I'd prefer it if we could use it instead of the preprocessor where possible. Any non-enabled `#ifdef` code will never hit the compiler after the preprocessing step, whereas anything with `if constexpr` will, and hence can be checked for correctness, even if not used.

Reply:
`if constexpr` works only for innerProd and innerProdC, but not for matrixVecProd and matrixMatProd. So, to be consistent, I switched using `#ifdef` for all places using BLAS operations. Let me know if you still think that using `if constexpr` is the better idea for innerProd and innerProdC?

Reply:
I think we should figure out how to keep `if constexpr` in here as it will be much safer and maintainable.

Reply:
In this case, using define directive to define undefined BLAS variables in matrixVecProd and matrixMatProd is a safe and practical approach to tackle this problem. What do you think?

Reply:
I still think enabling this with `if constexpr` is the way to go. Since half the code gets disabled before hitting the compiler, going with `if constexpr` allows for better static code analysis, as well as a feature of the modern uses of the language. Avoiding the preprocessor unless required has become a common theme in many C++ articles (https://cor3ntin.github.io/posts/undef_preprocessor/ is a good example). If using `if constexpr` is not possible, maybe in this instance many of the helpers in the `type_traits` library could help out here (https://en.cppreference.com/w/cpp/header/type_traits), such as `enable_if` for example.